CIP-1694/README.md

```diff
@@ -0,0 +1,504 @@
+
+Some of the parameters given in this table ( $P_1$ ... $R_{6c}$, $T_1$ ... $T_3$ ) may be updatable protocol parameters, but others should be hard-coded. This proposal deliberately leaves both this choice and the choice of actual parameter values open for discussion.
+
+> **Note**
+> For all treasury withdrawals, the withdrawal threshold is the **total** amount of Lovelace that is withdrawn by the action, not the amount of any single withdrawal if the action specifies more than one withdrawal.
+
+##### Governance Action Lifecycle
+
+Governance actions are checked for ratification only on an epoch boundary. This delay allows everyone to vote on each proposal and prove that they are active.
+
+**At most one** governance action **of each type** (i.e. one from each of the six categories listed above) can be staged for enactment in any given epoch. For example, there can only be one treasury withdrawal action in a single epoch (it may, however, comprise many individual withdrawals).
+
+Once ratified, actions will be staged for enactment. Actions that have been staged will be enacted on the following epoch boundary, unless they are dropped. All submitted governance actions will therefore either:
+
+1. be **ratified**;
+2. be **dropped** as a result of some higher priority action; or else
+3. will **expire** after a number of epochs.
+
+Deposits are returned immediately when either:
+
+1. a ratified action is **enacted**;
+1. the action **expires**; or
+1. a ratified action is **dropped**.
+
+Ratification is explained in more detail later in this document.
+
+>**Note:**
+>This design means that the first governance action of a given type to be ratified will be staged for enactment in an epoch.
+
+#### Enactment
+
+Ratified actions may be enacted at an epoch boundary. During enactment, actions in the staging group for the current epoch are prioritized as follows:
+
+1. Motion of no-confidence;
+2. New Constitutional Committee or a change to the quorum size;
+3. Updates to the Constitution;
+4. Hard Fork initiation;
+5. Protocol parameter changes;
+6. Treasury withdrawals.
+
+As described above, at most one action of each type may be enacted in any given epoch.
+So a ratified _"Motion of no-confidence"_ action will be enacted first if there is one, and so on.
+
+A successful _"Motion of no-confidence"_, the election of a new Constitutional Committee, or a constitutional change invalidates *all* other **unenacted** governance actions (whether or not they have been ratified),
+_causing them to be immediately **dropped** without ever being enacted_. Deposits for dropped actions will be returned immediately.
+
+#### Content of a Governance Action
+
+Every governance action will include the following:
+
+* a deposit amount (recorded since the amount is an updatable protocol parameter);
+* a reward address to receive the repaid deposit;
+* a URL to any metadata that is needed to justify the action;
+* a hash of the contents of this metadata URL.
```

Comment:
I'm unsure why we want to add another metadata mechanism for votes and DRep registration. I think we need a rationale section for this. I have two main questions:
1. Why is metadata important?
2. Given that we already have transaction metadata, and votes and registrations appear in metadata, if users _want_ to provide metadata about their votes or links to off-chain metadata, they can do that already using the transaction metadata in the transaction that contains the vote/certificate, no?

Reply:
It was very important to several people involved in the design to have a very high expectation that folks justify and provide context to all aspects of the governance. For example, the Constitutional Committee should justify _why_ their vote is in accordance with the constitution. The idea of the new metadata field is to make clear of this expectation. Also, one could have multiple votes and actions in a given transaction, so coupling things in the format removes the possibility of confusion in the usual metadata field (unless we added extra special validation to it, etc).

Note that the scheme above is what we do today for stake pool registrations: provide a URL for a JSON payload, and also the hash of the contents. The JSON payload contains a URL itself for a website.

But this absolutely needs a rationale in the rationale section. There are several folks who have valid concerns about this mechanism, and this is definitely a good thing for everyone to think about. What if the URL goes down, is the expectation that folks should not vote (or abstain)? Is this a DDOS vector again the governance system? In such a scenario, could folks just find another way to supply the hash pre-image? Should there be a summary of the justification in the JSON itself? Lots to think about!

Reply:
> The idea of the new metadata field is to make clear of this expectation.

Adding redundant fields to the permanent structure of transactions in order to socially encourage people to fill them in? That... doesn't seem like a good enough reason to me. I actually think this is one of the few _good_ uses of transaction metadata I've seen: it's _quite literally_ metadata about the transaction. I think this would need a really good justification for not using the field we already have when this is almost exactly it's purpose!

> Also, one could have multiple votes and actions in a given transaction, so coupling things in the format removes the possibility of confusion in the usual metadata field (unless we added extra special validation to it, etc).

We have standards for including structured data in transaction metadata, I don't think it would be hard to include a structured `voteIndex -> metadata url+hash` bit of data. I actually think this is one of the few _good_ uses of transaction metadata I've seen: it's _quite literally_ metadata about the transaction.

> Note that the scheme above is what we do today for stake pool registrations: provide a URL for a JSON payload, and also the hash of the contents. The JSON payload contains a URL itself for a website.

You will probably not be surprised to know that I've never liked that decision either :) 

> What if the URL goes down, is the expectation that folks should not vote (or abstain)? Is this a DDOS vector again the governance system? In such a scenario, could folks just find another way to supply the hash pre-image? Should there be a summary of the justification in the JSON itself? Lots to think about!

It seems to me like all of this should be _not the problem of layer 1_. People can wrangle out the details of this later, as they have very merrily been doing for other kinds of metadata. I don't see why we should commit to something like this as part of the core voting spec.